### DISCUSSION TITLE
In this version of Next.js iterating headers without awaiting first is still supported to facilitate migration but in a future version you will be required to await the result.

### DISCUSSION THREAD

**Discussion by kaijuh** [vercel/next.js#70633] [Help] [answered]
### Summary

I set up a new next app using nextauth and was able to set up a new google oath sign in and app without issues but I see this message on the console

`In route / headers were iterated implicitly with something likefor...of headers())or[...headers()], or explicitly with headers()[Symbol.iterator](https://github.com/nextauthjs/next-auth/issues/11929). headers()now returns a Promise and the return value should be awaited before attempting to iterate over headers. In this version of Next.js iterating headers without awaiting first is still supported to facilitate migration but in a future version you will be required to await the result. If thisheaders()` use is inside an async function await the return value before accessing attempting iteration. If this use is inside a synchronous function then convert the function to async or await the call from outside this function and pass the result `in.`

`In route / a cookie property was accessed directly with cookies().set('authjs.callback-url', ...). cookies() now returns a Promise and the return value should be awaited before accessing properties of the underlying cookies instance. In this version of Next.js direct access to cookies().set('authjs.callback-url', ...) is still supported to facilitate migration but in a future version you will be required to await the result. If this cookies() use is inside an async function await the return value before accessing attempting iteration. If this use is inside a synchronous function then convert the function to async or await the call from outside this function and pass the result in.`

`In route / a cookie property was accessed directly with cookies().set('authjs.pkce.code_verifier', ...). cookies() now returns a Promise and the return value should be awaited before accessing properties of the underlying cookies instance. In this version of Next.js direct access to cookies().set('authjs.pkce.code_verifier', ...) is still supported to facilitate migration but in a future version you will be required to await the result. If this cookies() use is inside an async function await the return value before accessing attempting iteration. If this use is inside a synchronous function then convert the function to async or await the call from outside this function and pass the result in.`

I searched for answers but I'm not sure how to address this issue. The app works as expected but it looks like it might break with future updates

Here are my dependencies

` "dependencies": {
    "@vercel/postgres": "^0.10.0",
    "next": "15.0.0-canary.173",
    "next-auth": "5.0.0-beta.22",
    "react": "^18",
    "react-dom": "^18"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "eslint": "^8",
    "eslint-config-next": "14.2.13",
    "typescript": "^5"
  }`

**Comment by icyJoseph** [comment, marked as answer]
I guess, next-auth uses those expressions. Looks like it is relatively new in Next.js, like 5 days old... according to:

- https://github.com/vercel/next.js/commit/05f159dffcd2e37837f9afb9e18927722147627a

So I would guess that next-auth, and other libraries are not yet updated to do the recommended approach.

Source of warnings in Next.js:

- https://github.com/vercel/next.js/blob/e5a0368a45c93afe53ad5535b3534923745e499d/packages/next/src/server/request/cookies.ts#L516
- https://github.com/vercel/next.js/blob/e5a0368a45c93afe53ad5535b3534923745e499d/packages/next/src/server/request/headers.ts#L431
- https://github.com/vercel/next.js/blob/cdb78b4e61a252ee9ed4105212809db42569b5f2/packages/next/src/server/request/draft-mode.ts#L169

Code in next-auth that breaks the above warnings:

- https://github.com/nextauthjs/next-auth/blob/0b94eab619d48a9a557d402567f4842d84ad188b/packages/next-auth/src/lib/actions.ts#L18

```js
import { headers as nextHeaders, cookies } from "next/headers"

// some lines later...
  const headers = new Headers(nextHeaders()) 
```

The same file contains the issue for `cookies`.  Give it some time and in the next versions of `next-auth` it will be fixed.

**Reply by kaijuh** [reply]
Thanks, I figured the issue was due to using Next Canary ("next": "15.0.0-canary.173") so I set up a new project using the stable version of Next JS ("next": "14.2.13",) yesterday and I got no warnings with the stable version of Next.

I am kind of on the fence because I was building a few projects using some of the latests NextCanary components so hopefully there won't be breaking changes. The 5 days old change looks about right when I started seeing those warnings in the console.

**Reply by icyJoseph** [reply]
Great, yeah, using RC has this kind of issue. Please don't forget to mark as answered :)
